### PR TITLE
stop watching resources when tree node is collapsed (#132)

### LINKED ISSUE #132: Should stop watching resources when tree node is collapsed
We currently start watching resources when a tree node is expanded. We should now also stop watching resources when a node is collapsed.

### PR DESCRIPTION
**Pull request by adietish**
fixes #132

### CODE REVIEW

**Comment by adietish** [comment]
in a first test this seems not to work: 
steps:
1. ASSERT: have a custom resource definition:
```
echo "apiVersion: apiextensions.k8s.io/v1beta1
kind: CustomResourceDefinition
metadata:
  name: cars.openshift.pub
spec:
  group: openshift.pub
  names:
    kind: Car
    listKind: CarList
    plural: cars
    singular: car
  scope: Cluster
  subresources:
    status: {}
  versions:
  - name: v1alpha1
    served: true
    storage: false
  - name: v1beta1
    served: true
    storage: true
  names:
    kind: Car
    plural: cars" | kubectl apply -f -
```
2. EXEC: unfold "Custom Resources" > "cars"
3. EXEC: collapse "cars"
4. EXEC: repeat 2 & 3 several times and eventually have "cars" unfolded
5. EXEC: create a new car: 
```
$ echo "apiVersion: openshift.pub/v1beta1
kind: Car
metadata:
  name: chevrolet
spec:
  date_of_manufacturing: "2002-07-01T00:00:00Z"
  engine: CH654321" | kubectl apply -f -
```

Result:
There's no child popping up for "Custom Resources" > "cars"

Expected Result:
"Custom Resources" > "cars" should have a child popping up called "chevrolet"

**Comment by adietish** [comment]
Works for me now. It can take a few sec before the new car appears though, because watching happens in the background, in a thread.

File: src/main/kotlin/com/redhat/devtools/intellij/kubernetes/model/ResourceModel.kt [hunk 1]
```diff
@@ -124,6 +126,14 @@ open class ResourceModel(
         contexts.current?.watch(definition)
     }
 
+    override fun stopWatch(kind: ResourceKind<out HasMetadata>) {
+        contexts.current?.watch(kind)
```
Comment (lstocchi) [inline src/main/kotlin/com/redhat/devtools/intellij/kubernetes/model/ResourceModel.kt:130]: Should this be stopWatch?

File: src/main/kotlin/com/redhat/devtools/intellij/kubernetes/model/ResourceModel.kt [hunk 2]
```diff
@@ -124,6 +126,14 @@ open class ResourceModel(
...
+    }
+
+    override fun stopWatch(definition: CustomResourceDefinition) {
+        contexts.current?.watch(definition)
```
Comment (lstocchi) [inline src/main/kotlin/com/redhat/devtools/intellij/kubernetes/model/ResourceModel.kt:134]: Should this be stopWatch?

**Review by lstocchi — CHANGES_REQUESTED** [review]

File: src/main/kotlin/com/redhat/devtools/intellij/kubernetes/model/ResourceModel.kt (hunk 1 above)
Comment (adietish) [reply src/main/kotlin/com/redhat/devtools/intellij/kubernetes/model/ResourceModel.kt:130]: ouch, thanks for pointing out.

**Comment by adietish** [comment]
@lstocchi verified, usecase works for me. Also further iterations where I collapse, unfold, create, delete etc. Please verify.

**Review by lstocchi — COMMENTED** [review]
`Stop watching` works fine but i have a doubt (bc i had complains for this in IJ tekton as well). If you remove the watch and something changes on cluster, when you re-expand the node you don't see the actual situation. 
BTW the behavior it is the same as in vscode with this. 

E.g I expand the tree until secrets 
image

I create a new secret
```
apiVersion: v1
kind: Secret
metadata:
  name: test-1
type: Opaque
data:
  password: b3Bz
```
The tree is updated fine. Now i collapse the secrets node, then the configuration node.
I delete some secrets, create a new one. 
At this point if i re-expand the secrets node i don't see the real situation.

**Comment by adietish** [comment]
> `Stop watching` works fine but i have a doubt (bc i had complains for this in IJ tekton as well). If you remove the watch and something changes on cluster, when you re-expand the node you don't see the actual situation.
gotcha, good catch!

I need to figure out why the following happens:

*Steps:*
1. ASSERT: have Configuration > Secrets expanded
2. EXEC: create secret:
```
echo "apiVersion: v1
kind: Secret
metadata:
  name: test-1
type: Opaque
data:
  password: b3Bz" | kubectl apply -f -
```
3. ASSERT: secret "test-1" appears
4. EXEC: collapse Secret, Configuration
5. EXEC: delete secret "test-1", create secret "test-2"
6. EXEC: expand Configuration > Secret

*Result:*
Secrets "test-1", "test-2" are listed

*Expected Result:**
Only secret "test-2" is listed.

**Comment by adietish** [comment]
The reason for this is that the resource was deleted when the watch wasn't watchig (was stopped). When the "Secrets" node is expanded, the cache still contains the secret that was killed. It looks as if fixing this is simple: clear the cache when stopping the watch.

**Comment by adietish** [comment]
@lstocchi fixed via clearing the cache when stopping the watch. Please re-review.

**Review by lstocchi — APPROVED** [review]
Awesome, works perfectly!! GJ!!

**Pull request merged**